gh-93162: Add ability to configure QueueHandler/QueueListener together and provide getHandlerByName() and getHandlerNames() APIs. (#93269)
python/cpython: Doc/library/logging.config.rst

You don't need to wrap with :func:`staticmethod` if you're setting the import
callable on a configurator *instance*.

.. _configure-queue:

Configuring QueueHandler and QueueListener
""""""""""""""""""""""""""""""""""""""""""

If you want to configure a :class:`~logging.handlers.QueueHandler`, noting that this
is normally used in conjunction with a :class:`~logging.handlers.QueueListener`, you
can configure both together. After the configuration, the ``QueueListener`` instance
will be available as the :attr:`listener` attribute of the created handler, and that
in turn will be available to you using :func:`~logging.getHandlerByName` and passing

> whatever name you have used for the ``QueueHandler`` in your configuration. The

Reviewer: ```suggestion
the name you have used for the ``QueueHandler`` in your configuration. The
```

dictionary schema for configuring the pair is shown in the example YAML snippet below.

.. code-block:: yaml

    handlers:
      qhand:
        class: logging.handlers.QueueHandler
        queue: my.module.queuefactory
        listener: my.package.CustomListener
        handlers:
          - hand_name_1
          - hand_name_2
          ...

The ``queue`` and ``listener`` keys are optional.

If the ``queue`` key is present, the corresponding value can be one of the following:

* An actual instance of :class:`queue.Queue` or a subclass thereof. This is of course
  only possible if you are constructing or modifying the configuration dictionary in
  code.

* A string that resolves to a callable which, when called with no arguments, returns
  the :class:`queue.Queue` instance to use. That callable could be a
  :class:`queue.Queue` subclass or a function which returns a suitable queue instance.

If the  ``queue`` key is absent, a standard unbounded :class:`queue.Queue` instance is
created and used.

> If the ``listener`` key is present, the corresponding value can be one of the following:

Reviewer: You say "one of the following" and then there is only one listed?

* A subclass of :class:`logging.handlers.QueueListener`. This is of course only
  possible if you are constructing or modifying the configuration dictionary in
  code.

If the ``listener`` key is absent, :class:`logging.handlers.QueueListener` is used.

The values under the ``handlers`` key are the names of other handlers in the
configuration (not shown in the above snippet) which will be passed to the queue
listener.

> Any custom queue handler and listener classes will need to deal with the same

Reviewer: ```suggestion
Any custom queue handler and listener classes will need be defined with the same
```

  code.

* A string that resolves to a callable which, when called with no arguments, returns
  the :class:`queue.Queue` instance to use. That callable could be a
  :class:`queue.Queue` subclass or a function which returns a suitable queue instance,
  such as ``my.module.queue_factory()``.

* A dict with a ``'()'`` key which is constructed in the usual way as discussed in
  :ref:`logging-config-dict-userdef`. The result of this construction should be a
  :class:`queue.Queue` instance.

If the  ``queue`` key is absent, a standard unbounded :class:`queue.Queue` instance is
created and used.

If the ``listener`` key is present, the corresponding value can be one of the following:

* A subclass of :class:`logging.handlers.QueueListener`. This is of course only
  possible if you are constructing or modifying the configuration dictionary in
  code.

* A string which resolves to a class which is a subclass of ``QueueListener``, such as
  ``'my.package.CustomListener'``.

* A dict with a ``'()'`` key which is constructed in the usual way as discussed in
  :ref:`logging-config-dict-userdef`. The result of this construction should be a

>   callable with the same signature as the ``QueueListener`` initializer.

Reviewer 2: ``The result of this construction should be a callable with the same signature...``

Does that mean ``()`` will be resolved, then called with the dict args and the result of that call should return a callable again?

Author: Yes. The callable will be called just as `QueueListener` would be.

Reviewer 2: But that's different than the normal behavior of '()':
````
In order to provide complete flexibility for user-defined object instantiation, the user needs to provide a ‘factory’ - 
a callable which is called with a configuration dictionary and which returns the instantiated object.
````
[from docs](https://docs.python.org/3/library/logging.config.html?highlight=dictconfig#user-defined-objects).
Note that the return normally is not a callable but an instantiated object.

Author: Yes, but in this case the value to be returned is documented explicitly. How can it be returned as an instantiated object in this case, if it has no access to the queue?

Author: Also note that "instantiated object" is a generic term. It does not imply any specific type of object, and a callable is also an instantiated object.

Reviewer 2: Do you think it would be possible to provide an example in the docs?
That'll make things more clear.

Author: Feel free to submit a documentation PR with what you think should be in there, and I'll take a look.
